tests/network/mod.rs

```diff
@@ -142,7 +183,165 @@ impl<D: DistAlgorithm> Adversary<D> for SilentAdversary {
     }
 }
 
+/// Return true with a certain `probability` ([0 .. 1.0])
+fn randomly(probability: f32) -> bool {
+    assert!(probability <= 1.0);
+    assert!(probability >= 0.0);
+
+    let mut rng = rand::thread_rng();
+    rng.gen_range(0.0, 1.0) <= probability
+}
+
+#[test]
+fn test_randomly() {
+    assert!(randomly(1.0));
+    assert!(!randomly(0.0));
+}
+
+/// An adversary that performs naive replay attacks
+///
+/// The adversary will randomly take a message that is sent to one of its nodes and re-send it to
+/// a different node
+#[allow(unused)] // not used in all tests
+pub struct RandomAdversary<D: DistAlgorithm, F> {
+    /// The underlying scheduler used
+    scheduler: MessageScheduler,
+
+    /// Node ids seen by the adversary.
+    known_node_ids: Vec<D::NodeUid>,
+    /// Node ids under control of adversary
+    known_adversarial_ids: Vec<D::NodeUid>,
+
+    /// Internal queue for messages to be returned on the next `Adversary::step()` call
+    outgoing: Vec<MessageWithSender<D>>,
+    /// Generates random messages to be injected
+    generator: F,
+
+    /// Probability of a message replay
+    p_replay: f32,
+    /// Probability of a message injection
+    p_inject: f32,
+}
+
+impl<D: DistAlgorithm, F> RandomAdversary<D, F> {
+    /// Creates a new random adversary instance
+    #[allow(unused)]
+    pub fn new(p_replay: f32, p_inject: f32, generator: F) -> RandomAdversary<D, F> {
+        assert!(
+            p_inject < 0.95,
+            "injections are repeated, p_inject must be smaller than 0.95"
+        );
+
+        RandomAdversary {
+            // the random adversary, true to its name, always schedules randomly
+            scheduler: MessageScheduler::Random,
+            known_node_ids: Vec::new(),
+            known_adversarial_ids: Vec::new(),
+            outgoing: Vec::new(),
+            generator,
+            p_replay,
+            p_inject,
+        }
+    }
+}
+
+impl<D: DistAlgorithm, F: Fn() -> TargetedMessage<D::Message, D::NodeUid>> Adversary<D>
+    for RandomAdversary<D, F>
+{
+    fn init(&mut self, nodes: &BTreeMap<D::NodeUid, Rc<NetworkInfo<D::NodeUid>>>) {
+        self.known_adversarial_ids = nodes.keys().cloned().collect();
+    }
+
+    fn pick_node(&mut self, nodes: &BTreeMap<D::NodeUid, TestNode<D>>) -> D::NodeUid {
+        // we are a bit hamstrung by the current API in that we would usually like a set of node
+        // ids available in `push_message`. the workaround is to "steal" them here for use later
+        if self.known_node_ids.is_empty() {
+            self.known_node_ids = nodes.keys().cloned().collect();
```

Comment:
It suffices to `panic` here if `known_node_ids` is empty because you should have already initialized it in `init`. `Adversary::pick_node` doesn't need to be self-mutating.